### PR TITLE
Auth redirect flow

### PR DESCRIPTION
**Pull request by raptisj**
## Type of change

- [x] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [x] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend-core`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/edge`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

## [WIP] 
This PR introduces an improved way to handle redirects coming from outside providers.
This is particularly useful when an OAuth connect flow fails. Instead of redirecting back to the profile with a closed modal(due to page refresh) it will persist state from a URL param and bring the page to its previous state(e.x. Open modal in the connect Oauth providers page).

### CODE REVIEW

File: packages/clerk-js/src/ui/Components.tsx [hunk 1]
```diff
@@ -149,7 +160,21 @@ const Components = (props: ComponentsProps) => {
   const { signInModal, signUpModal, userProfileModal, organizationProfileModal, createOrganizationModal, nodes } =
     state;
 
+  // TODO: rename these
+  const urlClerkState = getClerkQueryParam('__clerk_state') ?? '';
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/Components.tsx:164]: 🙃 Let's convert the `__clerk_state` into a constant and re-use it across the code base.

File: packages/clerk-js/src/ui/Components.tsx [hunk 2]
```diff
@@ -149,7 +160,21 @@ const Components = (props: ComponentsProps) => {
...
+  const decodedRedirectParams = urlClerkState ? JSON.parse(atob(urlClerkState)) : null;
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/Components.tsx:165]: 🔧 I'd suggest to move this to a separate utility helper e.g. `encodeBase64Json` and `decodeBase64Json`. Moreover, we must wrap this with a try/catch and cater for possible features and turn null on any error.

File: packages/clerk-js/src/ui/Components.tsx [hunk 3]
```diff
@@ -149,7 +160,21 @@ const Components = (props: ComponentsProps) => {
...
+
   useSafeLayoutEffect(() => {
+    if (urlClerkState && decodedRedirectParams && decodedRedirectParams?.modal) {
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/Components.tsx:168]:
```suggestion
    if (decodedRedirectParams && decodedRedirectParams?.modal) {
```

File: packages/clerk-js/src/ui/Components.tsx [hunk 4]
```diff
@@ -149,7 +160,21 @@ const Components = (props: ComponentsProps) => {
...
+      setState(s => ({
+        ...s,
+        [componentNodes[snakeToCamel(decodedRedirectParams.basePath)]]: {},
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/Components.tsx:171]: ❓ Component paths are dynamic. For example, the <UserProfile/> is located at `/user` by default but users can change it to anything. So the mapping between paths and components should be dynamic. 

Can we replace that basePath resolution with a property that says which component needs to be loaded on redirect? 

@nikosdouvlis WDTY?

File: packages/clerk-js/src/utils/getClerkQueryParam.ts [hunk 5]
```diff
@@ -20,3 +22,10 @@ export function getClerkQueryParam<T extends ClerkQueryParam>(param: T): ClerkQu
   const val = new URL(window.location.href).searchParams.get(param);
   return val ? (val as ClerkQueryParamsToValuesMap[T]) : null;
 }
+
+export function removeClerkQueryParam<T extends ClerkQueryParam>(param: T) {
```
Comment (SokratisVidros) [inline packages/clerk-js/src/utils/getClerkQueryParam.ts:26]: 🔧 Please add unit tests for this function.

File: packages/clerk-js/src/utils/getClerkQueryParam.ts [hunk 6]
```diff
@@ -20,3 +22,10 @@ export function getClerkQueryParam<T extends ClerkQueryParam>(param: T): ClerkQu
...
+  new URL(window.location.href).searchParams.delete(param);
+  const strippedUrl = new URL(`${window.location.origin}${window.location.pathname}`);
+  window.history.pushState({}, '', strippedUrl);
```
Comment (SokratisVidros) [inline packages/clerk-js/src/utils/getClerkQueryParam.ts:29]: 🔧 We need to make sure this solution will work across all routers and frameworks (e.g. Next.js, Remix, etc...). BTW, shouldn't we use replaceState instead of pushState?

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: packages/clerk-js/src/ui/Components.tsx (hunk 4 above)
Comment (nikosdouvlis) [reply packages/clerk-js/src/ui/Components.tsx:171]: A couple of things here.

1. As said before, reading + decoded + cleaning the state param should done with helper functions

2. We are only using this logic if the component was opened as a modal. So instead of always appending the state and checking the `modal` key, let's only append the state when it's needed in the first place. So if we have serialized state, we always perform the action dictated by the state (in this case, we open a modal)

3. We shouldn't have to do `options.startPath` . `options` is the public API of ClerkProvider and type changes to this are user-facing changes that we don't want 

4. I'm not sure what `componentNodes` represents or what `        [componentNodes[snakeToCamel(decodedRedirectParams.basePath)]]: {},` actually does... Can you please elaborate?

Comment (nikosdouvlis) [reply packages/clerk-js/src/ui/Components.tsx:171]: Generally, it might be easier to append the `componentName` and the current path when the flow starts and just append it to the startPath, eg: 
```
              startPath={'/organization' + serializedState.path}
```
(this is just an example as closing the modal afterward should also clear the path so next time it opens, it opens as it normally would)

File: packages/clerk-js/src/ui/components/UserProfile/ConnectedAccountsPage.tsx [hunk 7]
```diff
@@ -47,6 +47,15 @@ const AddConnectedAccount = () => {
   const user = useCoreUser();
   const { navigate } = useNavigate();
   const { strategies, strategyToDisplayData } = useEnabledThirdPartyProviders();
+  const { currentPath } = useRouter();
+  const isModal = useUserProfileContext().mode === 'modal';
+
+  const redirectParams = {
```
[the remark this answers is not in the document]
Comment (raptisj) [reply packages/clerk-js/src/ui/components/UserProfile/ConnectedAccountsPage.tsx:53]: @nikosdouvlis you are right, this is the case. So should we rename the url param from `__clerk_state` to `__clerk_modal_state`?

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+type SerializeAndAppendModalStateProps = { url: string; isModal: boolean; currentPath: string; componentName: string };
+export const serializeAndAppendModalState = ({
+  url,
+  isModal,
+  currentPath,
+  componentName,
+}: SerializeAndAppendModalStateProps) => {
+  if (!isModal) {
+    return url;
+  }
```
Comment (desiprisg) [inline packages/clerk-js/src/utils/queryStateParams.ts:26]: This has me a bit confused. Why are we passing `isModal` here if we are just going to return the `url`? Shouldn't this be handled outside the function? Am I missing something?

Comment (raptisj) [reply packages/clerk-js/src/utils/queryStateParams.ts]: @desiprisg I refactored the function as you suggested.

File: packages/clerk-js/src/ui/Components.tsx [hunk 9]
```diff
@@ -258,7 +276,7 @@ const Components = (props: ComponentsProps) => {
             <VirtualRouter
               preservedParams={PRESERVED_QUERYSTRING_PARAMS}
               onExternalNavigate={() => componentsControls.closeModal('userProfile')}
-              startPath='/user'
+              startPath={'/user' + urlStateParam?.path}
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/Components.tsx:279]: 🔧 Instead of string concatenation I'd suggest using the `buildUrl` helper that ensures proper & safer URL building.

File: packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx [hunk 10]
```diff
@@ -57,7 +57,7 @@ export const OrganizationProfileModal = (props: OrganizationProfileProps): JSX.E
   };
 
   return (
-    <Route path='organizationProfile'>
+    <Route path='organization'>
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx:60]: ❓ Is this change intentional? Won't this affect our current component routing?

File: packages/clerk-js/src/ui/elements/Accordion.tsx [hunk 11]
```diff
@@ -25,9 +25,15 @@ export const AccordionItem = (props: AccordionItemProps) => {
...
 
+  React.useEffect(() => {
+    setIsOpen(defaultOpen);
+  }, [defaultOpen]);
+
   React.useLayoutEffect(() => {
-    if (scrollOnOpen && isOpen && contentRef.current) {
-      contentRef.current.scrollIntoView({ behavior: 'smooth', block: 'center' });
+    if (scrollOnOpen && isOpen) {
+      window.requestAnimationFrame(() => {
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/elements/Accordion.tsx:34]: 🔧 Since `requestAnimationFrame` is used in a `useLayoutEffect` we should return a function with a cancelAnimationFrame to be 100% correct. PTAL https://css-tricks.com/using-requestanimationframe-with-react-hooks/

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+export const decodeBase64Json = (value: string) => {
+  return JSON.parse(atob(value));
+};
+
+export const encodeBase64Json = (value: { path: string; componentName: string }) => {
+  return btoa(JSON.stringify(value));
+};
+
+export const readAndRemoveStateParam = () => {
```
Comment (SokratisVidros) [inline packages/clerk-js/src/utils/queryStateParams.ts:11]: ❓ Does this function also remove the __clerk_state param from the URL or just reads it?

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    path: currentPath.replace(regexPattern, '') || '',
+    componentName,
+  };
+
+  const encodedRedirectParams = encodeBase64Json(redirectParams);
+
+  const urlWithParams = new URL(url);
+  const searchParams = urlWithParams.searchParams;
+
+  searchParams.set('__clerk_state', encodedRedirectParams);
```
Comment (SokratisVidros) [inline packages/clerk-js/src/utils/queryStateParams.ts:35]: ⛔ Let's rename the param to `__clerk_modal_state` as we already have a `__clerk_status` param for authentication purposes and the name will cause confusion to the future reader.

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+};
+
+export const readAndRemoveStateParam = () => {
+  const urlClerkState = getClerkQueryParam('__clerk_state') ?? '';
+
+  return urlClerkState ? decodeBase64Json(urlClerkState) : null;
+};
+
+type SerializeAndAppendModalStateProps = { url: string; currentPath: string; componentName: string };
+export const serializeAndAppendModalState = ({
```
Comment (SokratisVidros) [inline packages/clerk-js/src/utils/queryStateParams.ts:18]:
```suggestion
export const appendModalState = ({
```

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
+import { getClerkQueryParam } from '../utils';
+
+export const decodeBase64Json = (value: string) => {
```
Comment (SokratisVidros) [inline packages/clerk-js/src/utils/queryStateParams.ts:3]: These utilities exist in `encoders.ts`. Please fetch them from there.

File: packages/clerk-js/src/ui/hooks/useReadParamState.tsx [hunk 16]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
Comment (SokratisVidros) [inline packages/clerk-js/src/ui/hooks/useReadParamState.tsx:1]: 🔧 Please rename this to `useClerkModalState to dictate its usage

**Review by SokratisVidros — CHANGES_REQUESTED** [review]

File: packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx (hunk 10 above)
Comment (raptisj) [reply packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx:60]: @SokratisVidros this corresponds to the `startPath` props passed in VirtualRouter(Components.tsx). This change was made for consistency(we have `sign-in` not `signIn`)

cc: @nikosdouvlis

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  return urlClerkState ? JSON.parse(atob(urlClerkState)) : null;
+};
+
+type SerializeAndAppendModalStateProps = {
+  url: string;
+  startPath?: string;
+  currentPath?: string;
+  componentName: string;
+};
+export const appendModalState = ({
```
Comment (anagstef) [inline packages/clerk-js/src/utils/queryStateParams.ts:27]: 🙃  We could add tests for this.

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 18]
```diff
@@ -0,0 +1,51 @@
+import { encodeB64, getClerkQueryParam } from '../utils';
+
+export const buildUrl = ({ base, path }: { base: string; path: string | undefined }) => {
```
Comment (anagstef) [inline packages/clerk-js/src/utils/queryStateParams.ts:3]: 🔧 I see that we already have an exported `buildURL` from utils. We could rename this.

File: packages/clerk-js/src/ui/router/__tests__/VirtualRouter.test.tsx [hunk 19]
```diff
@@ -37,7 +37,7 @@ const ShowPreserved = () => {
 const Tester = () => (
   <VirtualRouter
     preservedParams={['preserved']}
-    startPath='/'
+    startPath={'/' + ''}
```
Comment (anagstef) [inline packages/clerk-js/src/ui/router/__tests__/VirtualRouter.test.tsx:40]: ❓ Is this for `tsc` I guess?

File: packages/clerk-js/src/ui/elements/NavigateToFlowStartButton.tsx [hunk 20]
```diff
@@ -11,6 +11,10 @@ export const useNavigateToFlowStart = () => {
     if (to !== router.currentPath) {
       return router.navigate(to);
     }
+
+    if (router.urlStateParam?.path) {
```
Comment (anagstef) [inline packages/clerk-js/src/ui/elements/NavigateToFlowStartButton.tsx:15]: 🙃 We can also add some tests for this to be safe

**Review by anagstef — CHANGES_REQUESTED** [review]
Looks good. Just the `buildUrl` naming needs to be refactored for me.

File: packages/clerk-js/src/utils/queryStateParams.ts (hunk 18 above)
Comment (raptisj) [reply packages/clerk-js/src/utils/queryStateParams.ts:3]: @anagstef the util you are referring to is doing another thing. It can't be used for our case and I think it should not be refactored to do many things.

File: packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx (hunk 10 above)
Comment (SokratisVidros) [reply packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx:60]: @nikosdouvlis ☝️

File: packages/clerk-js/src/utils/queryStateParams.ts (hunk 17 above)
Comment (raptisj) [reply packages/clerk-js/src/utils/queryStateParams.ts:28]: @anagstef added a test for this one

File: packages/clerk-js/src/utils/queryStateParams.ts (hunk 18 above)
Comment (raptisj) [reply packages/clerk-js/src/utils/queryStateParams.ts]: @anagstef changed the naming here

**Review by anagstef — APPROVED** [review]

File: packages/clerk-js/src/utils/queryStateParams.ts (hunk 15 above)
Comment (nikosdouvlis) [reply packages/clerk-js/src/utils/queryStateParams.ts:11]: This doesn't look like its used anywhere except for tests.
- Do we still need it?
- /utils exports an `encodeB64` maybe it also exports the corresponding decode method?
- If not, lets move it there and replace the JSON.parse inside readStateParam

File: packages/clerk-js/src/ui/components/CreateOrganization/CreateOrganization.tsx [hunk 21]
```diff
@@ -47,7 +47,7 @@ export const CreateOrganizationModal = (props: CreateOrganizationProps): JSX.Ele
   };
 
   return (
-    <Route path='createOrganization'>
+    <Route path='create-organization'>
```
Comment (nikosdouvlis) [inline packages/clerk-js/src/ui/components/CreateOrganization/CreateOrganization.tsx:50]: Why do we need this change?

File: packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx (hunk 10 above)
Comment (nikosdouvlis) [reply packages/clerk-js/src/ui/components/OrganizationProfile/OrganizationProfile.tsx:60]: I think this is safe, but Id rather  not include it in this PR if its not 100% required as the routing logic is pretty complex

File: packages/clerk-js/src/ui/components/CreateOrganization/CreateOrganization.tsx (hunk 21 above)
Comment (raptisj) [reply packages/clerk-js/src/ui/components/CreateOrganization/CreateOrganization.tsx:50]: @nikosdouvlis basically this is for convention reasons(see `sign-in`). Is this breaking? I don't think so

File: packages/clerk-js/src/utils/queryStateParams.ts (hunk 15 above)
Comment (raptisj) [reply packages/clerk-js/src/utils/queryStateParams.ts]: @nikosdouvlis I removed the the helper that was not used

File: packages/clerk-js/src/utils/queryStateParams.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+    componentName,
+    startPath,
+  };
+
+  const encodedRedirectParams = encodeB64(JSON.stringify(redirectParams));
+
+  const urlWithParams = new URL(url);
+  const searchParams = urlWithParams.searchParams;
+
+  searchParams.set('__clerk_modal_state', encodedRedirectParams);
```
Comment (SokratisVidros) [inline packages/clerk-js/src/utils/queryStateParams.ts:43]: Lets move the `__clerk_modal_state` to `core/constants.ts`

**Review by SokratisVidros — APPROVED** [review]

**Review by SokratisVidros — APPROVED** [review]

**Review by nikosdouvlis — APPROVED** [review]
